File: python/tests/shared.py
"""Helpers for the shared test cases of the repository's tests/ directory."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import gencmu
from gencmu._model import Token

REPOSITORY = Path(__file__).resolve().parents[2]
SHARED = REPOSITORY / "tests"


def cases(kind: str) -> list[Path]:
    return sorted((SHARED / kind).glob("*.json"))


def load_case(path: Path) -> dict[str, Any]:
    with open(path, encoding="utf-8") as file:
        return json.load(file)  # type: ignore[no-any-return]


def mismatch(pattern: Any, value: Any, where: str = "$") -> str | None:
    """Where a value fails to match a pattern (tests/README.md), or None."""
    if isinstance(pattern, dict):
        if not isinstance(value, dict):
            return f"{where}: expected an object, found {json.dumps(value, ensure_ascii=False)[:200]}"
        for key, expected in pattern.items():
            if key not in value:
                return f"{where}.{key}: missing"
            found = mismatch(expected, value[key], f"{where}.{key}")
            if found:
                return found
        return None
    if isinstance(pattern, list):
        if not isinstance(value, list) or len(value) != len(pattern):
            return f"{where}: expected {len(pattern)} elements, found {json.dumps(value, ensure_ascii=False)[:200]}"
        for index, (expected, found_value) in enumerate(zip(pattern, value)):
            found = mismatch(expected, found_value, f"{where}[{index}]")
            if found:
                return found
        return None
    if pattern != value or type(pattern) is not type(value):
        return f"{where}: expected {json.dumps(pattern, ensure_ascii=False)}, found {json.dumps(value, ensure_ascii=False)}"
    return None


def case_sources(case: dict[str, Any]) -> tuple[dict[str, str], str]:
    if "grammar" in case:
        grammar = case["grammar"]
        if "%ambiguity-resolution" not in grammar:
            grammar = "%ambiguity-resolution greedy ;\n" + grammar
        return (
            {
                "p.md": "## Main <?stage main?>\n\n- [main](main.md) <?grammar?>\n",
                "main.md": "```ebnf\n" + grammar + "\n```\n",
            },
            "p.md",
        )
    return dict(case["documents"]), case["pipeline"]


def case_tokens(case: dict[str, Any]) -> tuple[list[Token], str]:
    tokens: list[Token] = []
    position = 0
    for index, spec in enumerate(case["tokens"]):
        tags = {tag[1:] if tag.startswith("?") else tag: not tag.startswith("?") for tag in spec["tags"]}
        text = spec["text"]
        tokens.append(Token(text, tags, (index, index + 1), (position, position + len(text)), spec.get("phonemes")))
        position += len(text) + 1
    return tokens, " ".join(spec["text"] for spec in case["tokens"])


def run_case(case: dict[str, Any], use_cache: bool = True) -> tuple[dict[str, Any] | None, gencmu.ParseResult | None, gencmu.GencmuError | None]:
    sources, pipeline = case_sources(case)
    try:
        dialect = gencmu.load_dialect_sources(sources, pipeline, use_cache=use_cache)
    except gencmu.GencmuError as error:
        return None, None, error
    options = case.get("options", {})
    kwargs: dict[str, Any] = {
        "features": options.get("features", []),
        "auto_features": options.get("autoFeatures", False),
        "until": options.get("until"),
        "elision_only": options.get("elisionOnly"),
    }
    if "tokens" in case:
        tokens, text = case_tokens(case)
        result = dialect.parse_tokens(tokens, text, **kwargs)
    else:
        result = dialect.parse(case.get("input", ""), **kwargs)
    return gencmu.result_json(result), result, None

File: js/test/shared.js
// Runs the shared test cases of tests/, which every library runs.
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { Loader, resultJson, toBrackets, Token, GencmuError } from "../src/node.js";

export const repository = path.join(path.dirname(fileURLToPath(import.meta.url)), "..", "..");
export const grammars = path.join(repository, "grammars");

export function readGrammarFile(relative) {
  try {
    return fs.readFileSync(path.join(grammars, ...relative.split("/")), "utf8");
  } catch (error) {
    if (error.code === "ENOENT") return undefined;
    throw error;
  }
}

// A loader that reads the repository's grammars and, under `case/`, a
// case's own documents. One loader serves every case, since building one
// reads the Unicode table and the bootstrap, and its cache is keyed by each
// document's text as well as its path.
let caseDocuments = {};
const repositoryFiles = new Map();
const sharedLoader = new Loader((relative) => {
  if (relative.startsWith("case/")) return caseDocuments[relative.slice(5)];
  if (relative === "compiled.json") return undefined;
  if (!repositoryFiles.has(relative)) repositoryFiles.set(relative, readGrammarFile(relative));
  return repositoryFiles.get(relative);
});

export function loaderWith(documents) {
  caseDocuments = documents;
  return sharedLoader;
}

// Whether a value matches a pattern (tests/README.md).
export function matches(pattern, value) {
  if (Array.isArray(pattern)) {
    return Array.isArray(value) && pattern.length === value.length && pattern.every((item, index) => matches(item, value[index]));
  }
  if (pattern !== null && typeof pattern === "object") {
    return value !== null && typeof value === "object" && !Array.isArray(value) &&
      Object.keys(pattern).every((key) => matches(pattern[key], value[key]));
  }
  return pattern === value;
}

export function caseTokens(specs) {
  const tokens = [];
  let offset = 0;
  specs.forEach((spec, index) => {
    const tags = new Map();
    for (const tag of spec.tags) tags.set(tag.startsWith("?") ? tag.slice(1) : tag, !tag.startsWith("?"));
    const length = [...spec.text].length;
    tokens.push(new Token(tags, [index, index + 1], [offset, offset + length], spec.text, spec.phonemes ?? null, undefined));
    offset += length + 1;
  });
  return { tokens, text: specs.map((spec) => spec.text).join(" ") };
}

// Runs one engine case: the canonical result, the brackets and the error
// kind, or the grammar error.
export function runEngineCase(testCase) {
  let documents = testCase.documents || {};
  let pipeline = testCase.pipeline;
  if (testCase.grammar !== undefined) {
    const rules = testCase.grammar.includes("%ambiguity-resolution") ? testCase.grammar : `%ambiguity-resolution greedy ;\n${testCase.grammar}`;
    documents = { "main.md": "```ebnf\n" + rules + "\n```\n", "pipeline.md": "## Main <?stage main?>\n\n- [main](main.md) <?grammar?>\n" };
    pipeline = "pipeline.md";
  }
  let dialect;
  try {
    dialect = loaderWith(documents).dialect(`case/${pipeline}`);
  } catch (error) {
    if (error instanceof GencmuError) return { loadError: error };
    throw error;
  }
  const options = {
    features: new Set((testCase.options && testCase.options.features) || []),
    elisionOnly: testCase.options ? testCase.options.elisionOnly : undefined,
    autoFeatures: Boolean(testCase.options && testCase.options.autoFeatures),
  };
  let text = testCase.input;
  if (testCase.tokens) {
    const built = caseTokens(testCase.tokens);
    options.tokens = built.tokens;
    text = built.text;
  }
  const result = dialect.parse(text, options);
  return { result, json: resultJson(result), brackets: toBrackets(result) };
}
